HiPMobile/HiPMobile.Droid/Listeners/ExtendedLocationListener.cs

```diff
@@ -1,208 +1,384 @@
+                //time the location needs to be set
+            }
+            else
             {
-                alertDialog.Dispose(); // TODO this seems wrong
-            });
+                ShowSettingsAlert ();
+                Log.Info (LOG_TAG, "No location providers available");
+            }
+        }
 
-        // Showing Alert Message
-        alertDialog.Show();
-    }
+        #endregion
 
-        /// <summary>
-        /// Method that must be declared because of the inheritance of the Service class,
-        /// but always returns null.
-        /// </summary>
-        /// <param name="intent">Intent</param>
-        /// <returns>null</returns>
-        public override IBinder OnBind(Intent intent)
+        #region distance calculation
+
+        public double GetDistance (Location a, Location b)
         {
-            return null;
+            return DistanceCalculation (a.Latitude, a.Longitude, b.Latitude, b.Longitude);
         }
 
-        public void OnLocationChanged(Location location)
+        public double GetDistance (GeoLocation a, Location b)
         {
-           /* Activity activity = (Activity)ApplicationContext;
-            if (activity.GetType() == typeof(MainActivity)) {
-            MainActivity mainActivity = (MainActivity)activity;
-            //mainActivity.UpdatePosition(location);*/
+            return DistanceCalculation (a.Latitude, a.Longitude, b.Latitude, b.Longitude);
         }
 
+        public double GetDistance (Location a, GeoLocation b)
+        {
+            return DistanceCalculation (a.Latitude, a.Longitude, b.Latitude, b.Longitude);
+        }
 
-        public void OnProviderDisabled(string provider)
+        public double GetDistance (GeoLocation a, GeoLocation b)
         {
+            return DistanceCalculation (a.Latitude, a.Longitude, b.Latitude, b.Longitude);
         }
 
-        public void OnProviderEnabled(string provider)
+
+        //TODO: this is also implemented in the Exhibit.cs (nongenerated) code. 
+        //one implementation should suffice in the future (e.g. extra math/util class)
+        /// <summary>
+        ///     distance calculations as presented on http://andrew.hedges.name/experiments/haversine/
+        /// </summary>
+        //dlon = lon2 - lon1
+        //dlat = lat2 - lat1
+        //a = (sin(dlat/2))^2 + cos(lat1) * cos(lat2) * (sin(dlon/2))^2
+        //c = 2 * atan2(sqrt(a), sqrt(1-a) )
+        //d = R* c(where R is the radius of the Earth; 6373km as used in Exhibit class)
+        private double DistanceCalculation (double lat1, double long1, double lat2, double long2)
         {
+            var dlon = toRadians (long2 - long1);
+            var dlat = toRadians (lat2 - lat1);
+            var a = Math.Pow (Math.Sin (dlat / 2), 2) + Math.Cos (toRadians (lat1)) * Math.Cos (toRadians (lat2)) * Math.Pow (Math.Sin (dlon / 2), 2);
+            var c = 2 * Math.Atan2 (Math.Sqrt (a), Math.Sqrt (1 - a));
+            return 6373 * c;
         }
```

Comment:
Extract this to a Math helper class(in the PCL) from this class and the Exhibit.cs